core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/GroupNodeConstructor.java

```diff
@@ -7,17 +7,138 @@
 import ai.timefold.solver.core.api.score.Score;
 import ai.timefold.solver.core.api.score.stream.ConstraintStream;
 import ai.timefold.solver.core.config.solver.EnvironmentMode;
+import ai.timefold.solver.core.impl.util.Pair;
+import ai.timefold.solver.core.impl.util.Quadruple;
+import ai.timefold.solver.core.impl.util.Triple;
 
 public interface GroupNodeConstructor<Tuple_ extends AbstractTuple> {
 
-    static <Tuple_ extends AbstractTuple> GroupNodeConstructor<Tuple_>
-            of(NodeConstructorWithAccumulate<Tuple_> nodeConstructorWithAccumulate) {
-        return new GroupNodeConstructorWithAccumulate<>(nodeConstructorWithAccumulate);
+    static <CollectorA_, Tuple_ extends AbstractTuple> GroupNodeConstructor<Tuple_>
+            zeroKeysGroupBy(CollectorA_ collector, NoKeysOneCollectorGroupByBuilder<CollectorA_, Tuple_> builder) {
+        return new GroupNodeConstructorWithAccumulate<>(collector,
+                (groupStoreIndex, undoStoreIndex, nextNodesTupleLifecycle, outputStoreSize, environmentMode) -> builder.build(
+                        groupStoreIndex, undoStoreIndex, collector, nextNodesTupleLifecycle, outputStoreSize,
+                        environmentMode));
     }
 
-    static <Tuple_ extends AbstractTuple> GroupNodeConstructor<Tuple_>
-            of(NodeConstructorWithoutAccumulate<Tuple_> nodeConstructorWithoutAccumulate) {
-        return new GroupNodeConstructorWithoutAccumulate<>(nodeConstructorWithoutAccumulate);
+    static <CollectorA_, CollectorB_, Tuple_ extends AbstractTuple> GroupNodeConstructor<Tuple_>
+            zeroKeysGroupBy(CollectorA_ collectorA, CollectorB_ collectorB,
+                    NoKeysTwoCollectorsGroupByBuilder<CollectorA_, CollectorB_, Tuple_> builder) {
+        return new GroupNodeConstructorWithAccumulate<>(Pair.of(collectorA, collectorB),
+                (groupStoreIndex, undoStoreIndex, nextNodesTupleLifecycle, outputStoreSize, environmentMode) -> builder.build(
+                        groupStoreIndex, undoStoreIndex, collectorA, collectorB, nextNodesTupleLifecycle,
+                        outputStoreSize, environmentMode));
+    }
+
+    static <CollectorA_, CollectorB_, CollectorC_, Tuple_ extends AbstractTuple> GroupNodeConstructor<Tuple_>
+            zeroKeysGroupBy(CollectorA_ collectorA, CollectorB_ collectorB, CollectorC_ collectorC,
+                    NoKeysThreeCollectorsGroupByBuilder<CollectorA_, CollectorB_, CollectorC_, Tuple_> builder) {
+        return new GroupNodeConstructorWithAccumulate<>(Triple.of(collectorA, collectorB, collectorC),
+                (groupStoreIndex, undoStoreIndex, nextNodesTupleLifecycle, outputStoreSize, environmentMode) -> builder.build(
+                        groupStoreIndex, undoStoreIndex, collectorA, collectorB, collectorC,
+                        nextNodesTupleLifecycle, outputStoreSize, environmentMode));
+    }
```

Comment:
This confused me at first.

- `CollectorA_` et al. are not `...ConstraintCollector` subtypes. (Because that would cause further explosion in the number of interfaces you'd have to create for this.)
- `NoKeysThreeCollectorsGroupByBuilder` also doesn't reference constraint collectors.
- But then the builder itself is `Group...Node::new`, which *requires* a collector implementation. `CollectorA_` is not a collector subtype.

How does this compile?

It compiles because `Group...Node::new` is covariant with `build(int groupStoreIndex, int undoStoreIndex, CollectorA_ collectorA, CollectorB_ collectorB, CollectorC_ collectorC, TupleLifecycle<Tuple_> nextNodesTupleLifecycle, int outputStoreSize, EnvironmentMode environmentMode)`. The compiler is fooled and if the generic type in the end doesn't match the expected collector type, we'll get a cast error at runtime.

This is not at all obvious.